experimental/SelectPanel2: Add back button

siddharthkp opened the pull request:
- Closes https://github.com/github/primer/issues/3061
- Demo: https://primer-0f789cb9c7-13348165.drafts.github.io/storybook/?path=/story/drafts-components-selectpanel-examples--nested-selection
- Integration tests with dotcom: https://github.com/github/github/pull/313651

joshblack commented:
@siddharthkp for reviewing, should I be looking at the semantics of the nested selection story or should I be just looking at the back button behavior? Seeing the initial listbox was surprising so just wanted to double-check what I should be looking at in this PR 👀

siddharthkp commented:
> @siddharthkp for reviewing, should I be looking at the semantics of the nested selection story or should I be just looking at the back button behavior? 

Both are welcome!

> Seeing the initial listbox was surprising

I can go either way on this 🤔 

Repositories from the first panel still could be treated as "selected", with PRs inside them being displayed outside the panel.

Orrr we could treat the first panel/layer simply as filtering buttons that lead to the actual "options" that can be selected.

What do you think?

joshblack reviewed (commented):
Going through this, something I'm not sure about is the `listbox` being used for nested selection in this way. Something that is disorienting is that you make a selection in a listbox and then shift to a text input and you don't necessarily know why depending on the screen reader. This kind of relationship almost feels more like checkbox groups or a tree.

When going through it with voiceover, I noticed that sometimes the virtual cursor and the focused element got out of sync in selection 🤔 Not sure why but this may be unrelated to this PR. It also seems like hitting escape out of the second dialog will not restore focus to the button and will make it so that you cannot open the dialog again.

joshblack commented on packages/react/src/drafts/SelectPanel2/SelectPanel.examples.stories.tsx:
```diff
@@ -639,6 +655,131 @@ export const ShortSelectPanel = () => {
+    <>
+      <h1>Nested selection</h1>
+
+      <Button
+        ref={anchorRef}
+        onClick={() => setPanelToShow('repos')}
+        variant="invisible"
+        trailingAction={GearIcon}
+        sx={{width: '200px', '[data-component=buttonContent]': {justifyContent: 'start'}}}
+      >
```
Strike this, I see now the part where James mentioned this isn't needed!

siddharthkp commented:
> Going through this, something I'm not sure about is the listbox being used for nested selection in this way. Something that is disorienting is that you make a selection in a listbox and then shift to a text input and you don't necessarily know why depending on the screen reader. 

Very valid! 

I changed it to `role=list` for now and added a warning about accessibility on the story. Let's focus on shipping the back button and then we can get some guidance on the accessible markup for this story!
